### PR TITLE
Issue #7271 - Addressing MAX_CAPACITY limit in ArrayTernaryTrie and ArrayTrie

### PR DESCRIPTION
**Pull request by tinswzy**
#7271 
Signed-off-by: tinswzy <tinswzy@gmail.com>

### CODE REVIEW

**Comment by sbordet** [comment]
@lorban can you review this? Is the backporting fine, as in 9.4.x the implementation is different?

**Comment by ocean613** [comment]
@tinswzy  @lorban  When will this patch be available? I have a similar problem when the “_rows” variable overflows, and the ArrayTernaryTrie.getBest method returns an unexpected value.

**Comment by tinswzy** [comment]
@lorban @sbordet Any suggestions? Refactor  ArrayTernaryTrie/ArrayTire to support Integer.MAX_VALUE capacity or set the maximum Character.MAX_VALUE as before?

File: jetty-util/src/test/java/org/eclipse/jetty/util/TrieTest.java [hunk 1]
```diff
@@ -250,4 +254,60 @@ public void testFull(Trie<Integer> trie) throws Exception
...
+    {
+        ArrayTernaryTrie<String> trie = new ArrayTernaryTrie<>(Character.MAX_VALUE);
+
+        char[] c1 = new char[Character.MAX_VALUE - 1];
+        Arrays.fill(c1, 'a');
+        String huge = new String(c1);
+        assertTrue(trie.put(huge, "wow"));
+        assertThat(trie.get(huge), is("wow"));
+
+        assertThrows(IllegalArgumentException.class, () -> new ArrayTrie<String>(Character.MAX_VALUE + 1));
```
Comment (lorban) [inline jetty-util/src/test/java/org/eclipse/jetty/util/TrieTest.java]: There is already a separate test for this assertion, it can be removed.

File: jetty-util/src/test/java/org/eclipse/jetty/util/TrieTest.java [hunk 2]
```diff
@@ -250,4 +254,60 @@ public void testFull(Trie<Integer> trie) throws Exception
...
+    }
+
+    @Test
+    public void testArrayTernaryOverflowReject()
```
Comment (lorban) [inline jetty-util/src/test/java/org/eclipse/jetty/util/TrieTest.java]: This test name looks odd. Should it read `testArrayTrieOverflowReject()`?

Also, this test should also work for `ArrayTernaryTrie` so you should parameterize this test too.

File: jetty-util/src/test/java/org/eclipse/jetty/util/TrieTest.java [hunk 3]
```diff
@@ -250,4 +254,60 @@ public void testFull(Trie<Integer> trie) throws Exception
...
+    }
+
+    @Test
+    public void testArrayTernaryTrieCapacityOverFlow()
+    {
+        assertThrows(IllegalArgumentException.class, () -> new ArrayTernaryTrie<String>(Character.MAX_VALUE + 1));
+    }
+
+    @Test
+    public void testArrayTrieCapacity()
```
Comment (lorban) [inline jetty-util/src/test/java/org/eclipse/jetty/util/TrieTest.java]: This test and the one below (`testArrayTernaryTrieCapacity`) only differ by their trie implementation. You should refactor them as a single parameterized test.

File: jetty-util/src/test/java/org/eclipse/jetty/util/TrieTest.java [hunk 4]
```diff
@@ -250,4 +254,60 @@ public void testFull(Trie<Integer> trie) throws Exception
...
+        assertTrue(trie.put("", "/"));
+        assertThat(trie.getBest("/", 1, 0), is("/"));
+
+        char[] c1 = new char[Character.MAX_VALUE];
+        Arrays.fill(c1, 'a');
+        String huge = new String(c1);
+        assertFalse(trie.put(huge, "overflow"));
+        assertNull(trie.get(huge));
+
+        // The previous variable is not overwritten
```
Comment (lorban) [inline jetty-util/src/test/java/org/eclipse/jetty/util/TrieTest.java]: I guess you mean `The previous entry was not overridden`?

File: jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTrie.java [hunk 5]
```diff
@@ -49,6 +49,7 @@
  */
 public class ArrayTrie<V> extends AbstractTrie<V>
 {
+    public static int MAX_CAPACITY = Character.MAX_VALUE;
```
Comment (lorban) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTrie.java:52]: This should be `private` and `final`.

File: jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTernaryTrie.java [hunk 6]
```diff
@@ -530,7 +541,7 @@ public boolean isEmpty()
     @Override
     public boolean isFull()
     {
-        return _rows + 1 == _key.length;
+        return _rows + 1 >= _key.length;
```
Comment (lorban) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTernaryTrie.java:550]: This will overflow if `_rows == Character.MAX_VALUE`

File: jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTernaryTrie.java [hunk 7]
```diff
@@ -190,7 +201,7 @@ public boolean put(String s, V v)
                 if (t == _rows)
                 {
                     _rows++;
-                    if (_rows >= _key.length)
+                    if (_rows > _key.length)
```
Comment (lorban) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTernaryTrie.java]: This change will make line 209 throw `ArrayIndexOutOfBoundsException` if `_rows == _key.length` which happens when `t == Character.MAX_VALUE`.

**Review by lorban — CHANGES_REQUESTED** [review]
I'm not against the idea of backporting the `MAX_CAPACITY` check from the jetty-10.0.x branch, but the 10.0.x `ArrayTernaryTrie` seems to be buggy in that regard so I would not backport those changes.

The simplest thing to do would probably be to keep the 9.4.x code as-is, add the max capacity check with the `MAX_CAPACITY` at `Character.MAX_VALUE -1` but that value should be tested.

I'm also not certain `ArrayTrie` can support 65k entries so that would need to be tested too.

File: jetty-util/src/test/java/org/eclipse/jetty/util/TrieTest.java (hunk 1 above)
Comment (tinswzy) [reply jetty-util/src/test/java/org/eclipse/jetty/util/TrieTest.java]: Got it!

File: jetty-util/src/test/java/org/eclipse/jetty/util/TrieTest.java (hunk 2 above)
Comment (tinswzy) [reply jetty-util/src/test/java/org/eclipse/jetty/util/TrieTest.java]: Ok

File: jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTernaryTrie.java (hunk 7 above)
Comment (tinswzy) [reply jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTernaryTrie.java]: I will fix it~

**Comment by tinswzy** [comment]
> I'm not against the idea of backporting the `MAX_CAPACITY` check from the jetty-10.0.x branch, but the 10.0.x `ArrayTernaryTrie` seems to be buggy in that regard so I would not backport those changes.
> 
> The simplest thing to do would probably be to keep the 9.4.x code as-is, add the max capacity check with the `MAX_CAPACITY` at `Character.MAX_VALUE -1` but that value should be tested.
> 
> I'm also not certain `ArrayTrie` can support 65k entries so that would need to be tested too.

I think you're right. Just "Add the Max capacity check" is a good choice.

**Comment by lorban** [comment]
There still are problems in your PR; they were quicker to fix than to explain so I've created the following PR you're welcome to check and comment on: https://github.com/eclipse/jetty.project/pull/7405

In a nutshell, the fundamental capacity of the tries is 65534 entries as index values 0 and 65535 are special cases respectively in ArrayTrie and ArrayTernaryTrie. I've added the missing checks and made sure to not overflow `char`. I've also modified the test using the empty string as a key as this is a special case of the ArrayTrie implementation.

**Comment by joakime** [comment]
@lorban now that your PR #7405 is merged with a more comprehensive fix, can we close this one?

**Comment by lorban** [comment]
Yes, this issue can now be closed as it was superseded by https://github.com/eclipse/jetty.project/pull/7405.
